akka-remote/src/main/scala/akka/remote/artery/LruBoundedCache.scala

```diff
@@ -0,0 +1,234 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import akka.util.{ OptionVal, Unsafe }
+
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+object FastHash {
+
+  val Prime1: Int = 32717
+  val Prime2: Int = 32749
+  val Prime3: Int = 65519
+  val Prime4: Int = 65521
+
+  // FIXME: !!! Totally unscientific hash !!!, until I figure out how to port xxHash efficiently (it needs 64 bit access to be
+  // fast so we need the ByteBuffer instead of the String)
+  def ofString(s: String): Int = {
+    val chars = Unsafe.instance.getObject(s, EnvelopeBuffer.StringValueFieldOffset).asInstanceOf[Array[Char]]
+    var offset = 0
+    var remaining = chars.length
+
+    var v1 = Prime4
+    var v2 = Prime1
+    var v3 = Prime2
+    var v4 = Prime3
+
+    if (remaining >= 4) {
+      do {
+        // Do 4 parallel rounds (superscalar friendly)
+        v1 += java.lang.Integer.rotateLeft(chars(offset).toShort, 13) * Prime1
+        v2 += java.lang.Integer.rotateRight(chars(offset + 1).toShort, 13) * Prime2
+        v3 += java.lang.Integer.rotateLeft(chars(offset + 2).toShort, 15) * Prime3
+        v4 += java.lang.Integer.rotateRight(chars(offset + 3).toShort, 15) * Prime4
+        remaining -= 4
+        offset += 4
+      } while (remaining >= 4)
+    }
+
+    // Mix back parallel streams
+    var hash = v1 * v2 * v3 * v4 + chars.length
+
+    // add remaining chars
+    while (remaining > 0) {
+      hash *= chars(offset)
+      offset += 1
+      remaining -= 1
+    }
+
+    // Finalizer round similar to xxHash
+    hash ^= hash >> 15
+    hash *= Prime1
+    hash ^= hash >> 13
+    hash *= Prime2
+    hash ^= hash >> 16
+
+    hash
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] abstract class LruBoundedCache[K: ClassTag, V <: AnyRef: ClassTag](capacity: Int, evictAgeThreshold: Int) {
+  require(capacity > 0, "Capacity must be larger than zero")
+  require((capacity & (capacity - 1)) == 0, "Capacity must be power of two")
+  require(evictAgeThreshold <= capacity, "Age threshold must be less than capacity.")
+
+  private[this] val Mask = capacity - 1
+
+  // Practically guarantee an overflow
+  private[this] var epoch = Int.MaxValue - 4
+
+  private[this] val keys = Array.ofDim[K](capacity)
+  private[this] val values = Array.ofDim[V](capacity)
+  private[this] val hashes = Array.ofDim[Int](capacity)
+  private[this] val epochs = Array.fill[Int](capacity)(epoch - evictAgeThreshold) // Guarantee existing "values" are stale
+
+  final def get(k: K): Option[V] = {
```

Comment:
I only added this method for testing, but maybe it is useful, yes. I can add an explicit `evict` method, too.